java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java

```diff
@@ -0,0 +1,494 @@
+            states.addAll(sv.setConstraint(input, NullConstraint.NULL));
+            states.addAll(sv.setConstraint(input, NullConstraint.NOT_NULL));
+            return states;
+          }
+        }));
+
+      }
+    }
+    return startingStates;
+  }
+
+  private void handleBlockExit(ExplodedGraph.ProgramPoint programPosition) {
+    CFG.Block block = programPosition.block;
+    Tree terminator = block.terminator();
+    if (terminator != null) {
+      switch (terminator.kind()) {
+        case IF_STATEMENT:
+          handleBranch(block, ((IfStatementTree) terminator).condition());
+          return;
+        case CONDITIONAL_OR:
+        case CONDITIONAL_AND:
+          handleBranch(block, ((BinaryExpressionTree) terminator).leftOperand());
+          return;
+        case CONDITIONAL_EXPRESSION:
+          handleBranch(block, ((ConditionalExpressionTree) terminator).condition());
+          return;
+        case FOR_STATEMENT:
+          ForStatementTree forStatement = (ForStatementTree) terminator;
+          ExpressionTree condition = forStatement.condition();
+          if (condition != null) {
+            handleBranch(block, condition, false);
+            return;
+          }
+          break;
+        case WHILE_STATEMENT:
+          ExpressionTree whileCondition = ((WhileStatementTree) terminator).condition();
+          handleBranch(block, whileCondition, !whileCondition.is(Tree.Kind.BOOLEAN_LITERAL));
+          return;
+        case SYNCHRONIZED_STATEMENT:
+          resetFieldValues();
+          break;
+        default:
+          // do nothing by default.
+      }
+    }
+    // unconditional jumps, for-statement, switch-statement, synchronized:
+    for (CFG.Block successor : block.successors()) {
+      enqueue(new ExplodedGraph.ProgramPoint(successor, 0), programState);
+    }
+  }
+
+  private void handleBranch(CFG.Block programPosition, Tree condition) {
+    handleBranch(programPosition, condition, true);
+  }
+
+  private void handleBranch(CFG.Block programPosition, Tree condition, boolean checkPath) {
+    Pair<List<ProgramState>, List<ProgramState>> pair = constraintManager.assumeDual(programState);
+    for (ProgramState state : pair.a) {
+      // enqueue false-branch, if feasible
+      ProgramState ps = ProgramState.stackValue(state, SymbolicValue.FALSE_LITERAL);
+      enqueue(new ExplodedGraph.ProgramPoint(programPosition.falseBlock(), 0), ps);
+      if (checkPath) {
+        alwaysTrueOrFalseChecker.evaluatedToFalse(condition);
+      }
+    }
+    for (ProgramState state : pair.b) {
+      ProgramState ps = ProgramState.stackValue(state, SymbolicValue.TRUE_LITERAL);
+      // enqueue true-branch, if feasible
+      enqueue(new ExplodedGraph.ProgramPoint(programPosition.trueBlock(), 0), ps);
+      if (checkPath) {
+        alwaysTrueOrFalseChecker.evaluatedToTrue(condition);
+      }
+    }
+  }
+
+  private void visit(Tree tree, @Nullable Tree terminator) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "visit" is 64 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) This method has 163 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS138)